### PR TITLE
Add GPT-2 Model and its Variants

### LINKED ISSUE #337: Investigate recreating a GPT-2 forward pass with KerasNLP
With the addition of a roughly compatible tokenizer with gpt-2 in https://github.com/keras-team/keras-nlp/pull/303, it should be possible to start experimenting with recreating an entire gpt forward pass with KerasNLP components.

This should be more an investigation and validation for now, as we continue to answer some core design questions with our BERT modeling code.

Rough flow:

 - Load a reference version of the gpt-2 tokenizer and network, either from the [released github](https://github.com/openai/gpt-2) or huggingface.
 - Recreate the gpt-2 tokenizer using https://github.com/keras-team/keras-nlp/pull/303
 - Recreate the decoder models itself using KerasNLP components, and convert weights.
 - Verify that our output can match with a high degree of accuracy.

### PR DESCRIPTION
**Pull request by abheesht17**
Partially resolves #337

### CODE REVIEW

File: keras_nlp/models/gpt2.py [hunk 1]
```diff
@@ -0,0 +1,287 @@
...
+        token_ids = keras.Input(shape=(None,), dtype="int32", name="token_ids")
+        padding_mask = keras.Input(
+            shape=(None,), dtype="int32", name="padding_mask"
+        )
+
+        # Embed tokens, positions.
+        token_embedding = keras.layers.Embedding(
+            input_dim=vocabulary_size,
+            output_dim=hidden_dim,
+            embeddings_initializer=_gpt_2_kernel_initializer(stddev=0.01),
```
Comment (mattdangerw) [inline keras_nlp/models/gpt2.py:107]: Awesome work grabbing the original initializers here. This is nice to have.

**Comment by mattdangerw** [comment]
This looks great! Will take a deeper look, but you can do two things for now

1) Comment with a pointer to wherever you are grabbing the hyperparameters from for the custom sizes. Does openai provide this anywhere directly?
2) Update your colab from the issue to actually use a gpt-2 symbols from this branch? That would be a nice validation that we have fully recreated the correct graph here.

File: keras_nlp/models/gpt2.py [hunk 2]
```diff
@@ -0,0 +1,287 @@
...
+        hidden_dim=1280,
+        intermediate_dim=5120,
+        dropout=0.1,
+        max_sequence_length=1024,
+        name=name,
+        trainable=trainable,
+    )
+
+
+def Gpt2XLarge(vocabulary_size, name=None, trainable=True):
```
Comment (mattdangerw) [inline keras_nlp/models/gpt2.py:254]: I think Gpt2ExtraLarge might actually fit better with our normal naming recommendations.

File: keras_nlp/models/gpt2.py (hunk 1 above)
Comment (abheesht17) [reply keras_nlp/models/gpt2.py:107]: Thanks! Posting the reference here, just to keep a record:

For embedding layers: https://github.com/openai/gpt-2/blob/master/src/model.py#L152-L155
For FF layers: https://github.com/openai/gpt-2/blob/master/src/model.py#L50

**Comment by abheesht17** [comment]
> This looks great! Will take a deeper look, but you can do two things for now
> 
> 1. Comment with a pointer to wherever you are grabbing the hyperparameters from for the custom sizes. Does openai provide this anywhere directly?
> 2. Update your colab from the issue to actually use a gpt-2 symbols from this branch? That would be a nice validation that we have fully recreated the correct graph here.

1. They have a config file in their checkpoint file named `hparams.json`. This is the format of the links: 
`https://openaipublic.blob.core.windows.net/gpt-2/models/{num_params}/hparams.json`.

If you want to double check once, please visit these links:
GPT-2 Small: https://openaipublic.blob.core.windows.net/gpt-2/models/124M/hparams.json
GPT-2 Medium: https://openaipublic.blob.core.windows.net/gpt-2/models/355M/hparams.json
GPT-2 Large: https://openaipublic.blob.core.windows.net/gpt-2/models/774M/hparams.json
GPT-2 ExtraLarge: https://openaipublic.blob.core.windows.net/gpt-2/models/1558M/hparams.json

2. Sure, will do this now!

File: keras_nlp/models/gpt2.py [hunk 3]
```diff
@@ -0,0 +1,287 @@
...
+from keras_nlp.layers import PositionEmbedding
+from keras_nlp.layers import TransformerDecoder
+
+
+def _gpt_2_kernel_initializer(stddev=0.02):
+    return keras.initializers.RandomNormal(stddev=stddev)
+
+
+class Gpt2Custom(keras.Model):
+    """Generative Pretrained Transformer-2 (GPT-2) network.
```
Comment (mattdangerw) [inline keras_nlp/models/gpt2.py:29]: I think we should get out of the habit of listing the full model name in a first line. That should be short, scannable, and informative.

`GPT-2 core network with custom hyperparmeters.` or something like that

File: keras_nlp/models/gpt2.py [hunk 4]
```diff
@@ -0,0 +1,287 @@
...
+                "hidden_dim": self.hidden_dim,
+                "intermediate_dim": self.intermediate_dim,
+                "dropout": self.dropout,
+                "max_sequence_length": self.max_sequence_length,
+            }
+        )
+        return config
+
+
+MODEL_DOCSTRING = """GPT-2 implementation using "{type}"
```
Comment (mattdangerw) [inline keras_nlp/models/gpt2.py:181]: Same here `GPT-2 core network with "{type}" architecture.`

File: keras_nlp/models/gpt2.py [hunk 5]
```diff
@@ -0,0 +1,287 @@
...
+        ),
+        "padding_mask": tf.constant([1] * 1024, shape=(1, 1024)),
+    }}
+    output = model(input_data)
+"""
+
+
+def Gpt2Base(vocabulary_size, name=None, trainable=True):
+    return Gpt2Custom(
+        vocabulary_size=vocabulary_size,
```
Comment (mattdangerw) [inline keras_nlp/models/gpt2.py:214]: Is `vocabulary_size` actually fixed for all released checkpoints? I think so, but wanted to check my understanding

File: keras_nlp/models/gpt2.py [hunk 6]
```diff
@@ -0,0 +1,287 @@
...
+
+        # Embed tokens, positions.
+        token_embedding = keras.layers.Embedding(
+            input_dim=vocabulary_size,
+            output_dim=hidden_dim,
+            embeddings_initializer=_gpt_2_kernel_initializer(stddev=0.01),
+            name="token_embedding",
+        )(token_ids)
+
+        position_embedding = PositionEmbedding(
```
Comment (mattdangerw) [inline keras_nlp/models/gpt2.py:111]: Maybe drop a comment that we can't use TokenAndPositionEmbedding because of the separate initializers for the random graph.

**Review by mattdangerw — CHANGES_REQUESTED** [review]
Overall lgtm, just a few minor comments.

**Review by jbischof — APPROVED** [review]
Great work, thank you!

One request: please move your style changes of other files to a separate PR.

**Comment by mattdangerw** [comment]
Also just a heads up I will be pushing a name change cleanup that will probably require updating any checkpoints conversion scripts using `TransformerDecoder`. But it's just reworking the layer names.

https://github.com/keras-team/keras-nlp/pull/353

File: keras_nlp/models/gpt2.py (hunk 3 above)
Comment (abheesht17) [reply keras_nlp/models/gpt2.py:29]: Cool! I have shifted the full name below

File: keras_nlp/models/gpt2.py [hunk 7]
```diff
@@ -0,0 +1,287 @@
...
+                "intermediate_dim": self.intermediate_dim,
+                "dropout": self.dropout,
+                "max_sequence_length": self.max_sequence_length,
+            }
+        )
+        return config
+
+
+MODEL_DOCSTRING = """GPT-2 implementation using "{type}"
+    architecture (with {num_params} parameters).
```
[the remark this answers is not in the document]
Comment (abheesht17) [reply keras_nlp/models/gpt2.py:182]: Changed it to `GPT-2 core network with "{type}" architecture ({num_params} parameters).` after addressing Matt's comment. Two "with"s might look weird in the same sentence

File: keras_nlp/models/gpt2.py (hunk 5 above)
Comment (abheesht17) [reply keras_nlp/models/gpt2.py:214]: Yep, fixed for all ckpts - 50257

https://openaipublic.blob.core.windows.net/gpt-2/models/355M/hparams.json, https://openaipublic.blob.core.windows.net/gpt-2/models/124M/hparams.json, etc.

File: keras_nlp/models/gpt2.py [hunk 8]
```diff
@@ -0,0 +1,287 @@
...
+        hidden_dim=1280,
+        intermediate_dim=5120,
+        dropout=0.1,
+        max_sequence_length=1024,
+        name=name,
+        trainable=trainable,
+    )
+
+
+def Gpt2ExtraLarge(vocabulary_size, name=None, trainable=True):
```
[the remark this answers is not in the document]
Comment (abheesht17) [reply keras_nlp/models/gpt2.py:254]: Yep, ckpts for all 4 sizes are present

**Comment by abheesht17** [comment]
> Also just a heads up I will be pushing a name change cleanup that will probably require updating any checkpoints conversion scripts using `TransformerDecoder`. But it's just reworking the layer names.

@mattdangerw, updated: https://colab.research.google.com/drive/1vI3dDmFEFOQM0GbDBfVf6sanATjkwV9X?usp=sharing!

File: keras_nlp/models/gpt2.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (abheesht17) [reply keras_nlp/models/gpt2.py]: On second thoughts, changed it to 

`GPT-2 "{type}" architecture with {num_params} parameters.`

[the remark this answers is not in the document]
Comment (mattdangerw) [reply keras_nlp/models/gpt2.py]: I really like the parameter count as a usability improvement, but I'm not sure we can do this totally accurately right?

vocabulary_size=10 vs vocabulary_size=1_000_000 would kinda invalidate your parameter count. I think we need to do this at the checkpoint level. Or say "roughly XXX parameters".

[the remark this answers is not in the document]
Comment (mattdangerw) [reply keras_nlp/models/gpt2.py]: Let's just remove this for now, and plan to document parameter count with actual checkpoints.

[the remark this answers is not in the document]
Comment (abheesht17) [reply keras_nlp/models/gpt2.py]: @mattdangerw, the reason why I put #params in the description is because that is how the official repo distinguishes between the different GPT-2 sizes: https://github.com/openai/gpt-2/blob/master/model_card.md. Even the paper does not mention "Base", "Medium", "Large", "ExtraLarge". However, since changing the vocabulary_size will change #param, I'll remove #param altogether, makes sense to have it at the ckpt level.

[the remark this answers is not in the document]
Comment (mattdangerw) [reply keras_nlp/models/gpt2.py]: Yeah totally. I am torn because it's very useful.

But GPT which is only releasing a single checkpoint so didn't have to deal with the general case here. And the parameter count difference between checkpoints can get pretty big.

Funnily enough, the original bert repo is actually falling into this trap. It is claiming the multi lingual version is 110M params when its almost 180. There's even a poor forgotten PR for it https://github.com/google-research/bert/pull/1150

**Comment by abheesht17** [comment]
Please hold off from merging this; I'll validate ckpt conversion for all 4 sizes

**Comment by abheesht17** [comment]
> Please hold off from merging this; I'll validate ckpt conversion for all 4 sizes

Good to go! :)
